# HW4

<b>Instructions:</b> Write your answers in the empty code cells below.

<b>Penalties:</b> You will incur penalties if:
<ul>
<li>Your code is wrong</li>
<li>Your code does not work on different data</li>
<li>Your code is unnecessarily slow (you use a for loop, or you use DataFrame.apply unnecessarily, etc)
</ul>

Start by downloading the file <i>cleaned_survey.csv</i> in the same folder as this .ipynb file. The file contains the results from the student survey.  

Then, execute the code below to load the data in a  dataframe <i>df</i>. 

In [1]:
import pandas as pd
df = pd.read_csv("cleaned_survey-1.csv", index_col=0)
df.drop(['Expert'],axis=1,inplace=True)
#df.drop(['Languages','Expert'],axis=1,inplace=True)

In [2]:
pd.set_option('display.max_columns',50)

In [3]:
df.head()

Job Program  ProgSkills  C  CPP   CS  Java  Python   JS    R  SQL  SAS  \
0  0.0    MSIS           4  1    1  0.0     0     1.0  1.0  0.0  1.0  0.0   
1  0.5    MSIS           3  1    1  0.0     1     0.0  0.0  0.0  1.0  0.0   
2  0.0    MSIS           3  0    0  0.0     1     1.0  0.0  0.0  1.0  0.0   
3  0.0    MSIS           3  1    0  0.0     1     1.0  0.0  1.0  1.0  0.0   
4  0.0    MSIS           3  1    0  0.0     1     1.0  0.0  0.0  1.0  0.0   

   Excel  Tableau  Regression  Classification  Clustering  Bach_0to1  \
0      1        1         1.0               4           4          0   
1      1        0         0.0               2           2          0   
2      1        0         1.0               3           3          0   
3      1        0         1.0               2           3          0   
4      1        0         0.0               1           1          0   

   Bach_1to3  Bach_3to5  Bach_5Plus  Languages  
0          1          0           0          6  
1          0          0           1          4  
2          0          1           0          3  
3          0          0           1          5  
4          0          1           0          4

### Data Set description

The data set contains one row for every person who filled the survey. We use <b>person</b> and <b>student</b> interchangeably.

<ul>
<li><b>Job</b>: 0 for those students without a job, 0.5 for those students with a part-time job, and 1 for those students with a full-time job
<li><b>Program</b>: program of enrollment
<li><b>C-Regression</b>: indicates whether the student knows (1) or doesn't know (0) a certain programming language or topic
<li><b>Classification</b>: indicates the level of knowledge (1-5) on classification
<li><b>Clustering</b>: indicates the level of knowledge (1-5) on clustering
<li><b>Bach_0to1</b>: 1 if time elapsed from graduation is less than a year; 0 otherwise
<li><b>Bach_1to3</b>: 1 if time elapsed from graduation is between one and three years; 0 otherwise
<li><b>Bach_3to5</b>: 1 if time elapsed from graduation is between three and five years; 0 otherwise
<li><b>Bach_5Plus</b>: 1 if time elapsed from graduation is more than five years; 0 otherwise
</ul>

### Question 1

<p>For each programming skill level, compute:</p>
<ol>
<li><b>PropPJC2</b>:the proportion (0 to 1, computed relatively to the group size) of people that know at least one among Python and Java and their Classification knowledge has to be at least 2. 
<li><b>StdevClust</b>:the standard deviation of the Clustering knowledge (function <i>std</i>)
</ol>
<p>You should return one data frame with one row for each programming skill level and two columns named <b>PropPy_or_Java_and_Class2+</b> and <b>StdevClass</b>. Don't worry about having a Hierarchical Index for the columns.</p> 

(In Camino, pick the correct **PropPy_or_Java_and_Class2+** number for student with ProgSkills = 5 )

In [4]:
df['JavaPython1'] = (df.Java==1)|(df.Python==1) #Creates new row JavaPython1 that returns true of students know
#at least one of these languages

In [5]:
df.JavaPython1 = df.JavaPython1.astype(int)  #converts boolean values in that column to 1s and 0s

In [6]:
df.head(8) #Looks good

Job                        Program  ProgSkills  C  CPP   CS  Java  Python  \
0  0.0                           MSIS           4  1    1  0.0     0     1.0   
1  0.5                           MSIS           3  1    1  0.0     1     0.0   
2  0.0                           MSIS           3  0    0  0.0     1     1.0   
3  0.0                           MSIS           3  1    0  0.0     1     1.0   
4  0.0                           MSIS           3  1    0  0.0     1     1.0   
5  1.0  Supply Chain Mgmt & Analytics           1  0    0  0.0     0     0.0   
6  0.0                           MSIS           3  1    1  0.0     1     0.0   
7  0.0                           MSIS           2  1    0  0.0     1     0.0   

    JS    R  SQL  SAS  Excel  Tableau  Regression  Classification  Clustering  \
0  1.0  0.0  1.0  0.0      1        1         1.0               4           4   
1  0.0  0.0  1.0  0.0      1        0         0.0               2           2   
2  0.0  0.0  1.0  0.0      1        0         1.0               3           3   
3  0.0  1.0  1.0  0.0      1        0         1.0               2           3   
4  0.0  0.0  1.0  0.0      1        0         0.0               1           1   
5  0.0  0.0  1.0  0.0      1        0         1.0               1           1   
6  0.0  0.0  1.0  0.0      1        0         0.0               2           2   
7  0.0  0.0  1.0  0.0      1        1         0.0               2           2   

   Bach_0to1  Bach_1to3  Bach_3to5  Bach_5Plus  Languages  JavaPython1  
0          0          1          0           0          6            1  
1          0          0          0           1          4            1  
2          0          0          1           0          3            1  
3          0          0          0           1          5            1  
4          0          0          1           0          4            1  
5          0          1          0           0          1            0  
6          0          1          0           0          4            1  
7          1          0          0           0          3            1

In [7]:
df[df.Classification>=2].groupby('ProgSkills')['JavaPython1'].sum()

ProgSkills
1     0
2     4
3    19
4     7
5     1
Name: JavaPython1, dtype: int64

This represents the number of students in each category who know Java **OR** Python, **AND** students who have a classification skill of 2 or greater.

In [8]:
df.groupby('ProgSkills')['JavaPython1'].count()

ProgSkills
1     7
2    11
3    29
4    12
5     2
Name: JavaPython1, dtype: int64

This represents the total amount of people (group size) in each category **whether or not** they know Java/Python.

Therefore we now must calculate the proportion of the two.

In [9]:
df[df.Classification>=2].groupby('ProgSkills')['JavaPython1'].sum()/df.groupby('ProgSkills')['JavaPython1'].count()

ProgSkills
1    0.000000
2    0.363636
3    0.655172
4    0.583333
5    0.500000
Name: JavaPython1, dtype: float64

This looks good, let's assign this to a variable.

In [10]:
jpython=\
df[df.Classification>=2].groupby('ProgSkills')['JavaPython1'].sum()/df.groupby('ProgSkills')['JavaPython1'].count()

In [11]:
df.groupby('ProgSkills')['Clustering'].std() #std dev of clustering knowledge for each programming skill level

ProgSkills
1    0.786796
2    0.687552
3    1.085053
4    1.083625
5    0.707107
Name: Clustering, dtype: float64

In [12]:
cluster_sd = df.groupby('ProgSkills')['Clustering'].std() #Lets store this in a dataframe

In [13]:
dfboth = pd.concat([jpython, cluster_sd],axis=1).\
rename(columns={'JavaPython1':'PropPy_or_Java_and_Class2+','Clustering':'StdevClass'})

In [14]:
dfboth

PropPy_or_Java_and_Class2+  StdevClass
ProgSkills                                        
1                             0.000000    0.786796
2                             0.363636    0.687552
3                             0.655172    1.085053
4                             0.583333    1.083625
5                             0.500000    0.707107

Above, we concatonated both dataframes into one showing proportion and standard deviation for each programming skill level.

### Question 2

For each program, how many people have a Programming Skill knowledge of less than 4? Report one row for each existing program. 

(In Camino, fill in the number of students for MBA and MSIS in a format of MBA_total/MSIS_total. For example : 10/23 )

In [15]:
df[df.ProgSkills<4].groupby('Program')['ProgSkills'].count()

Program
Business Man                      1
Faculty!                          1
MBA                              13
MSIS                             30
Supply Chain Mgmt & Analytics     2
Name: ProgSkills, dtype: int64

The above groupby statement returns the number of people in each program with programming skills less than four. MBA and MSIS students have 13 and 30 students respectively who satisfy those constraints. 

### Question 3

For each Job, return the rows of those people who, within their group, know the most Languages. (In Camino, pick the correct total number of people who satisfy this condition)

In [16]:
df.groupby('Job')['Languages'].max()

Job
0.0    7
0.5    6
1.0    7
Name: Languages, dtype: int64

Above, we have the maximum number of values each group possesses. For example, within the part-time category (0.5); the highest number of languages known among anyone in that category is 6. 

To make this work for any program with any data, let's store the maximum value within each category in a variable.

In [17]:
no_job = df.groupby('Job')['Languages'].max()[0]   #stores score of 7

In [18]:
part_time = df.groupby('Job')['Languages'].max()[0.5]   #stores score of 6

In [19]:
full_time = df.groupby('Job')['Languages'].max()[1]    #stores score of 7

Now let's make a boolean to yield the desired results, using our new variables.

In [20]:
df[((df.Languages == no_job)&(df.Job==0.0))|
   ((df.Languages == part_time)&(df.Job==0.5))|((df.Languages == full_time)&(df.Job==1.0))]

Job Program  ProgSkills  C  CPP   CS  Java  Python   JS    R  SQL  SAS  \
23  0.5    MSIS           3  1    1  0.0     1     1.0  1.0  0.0  1.0  0.0   
45  1.0    MSIS           3  1    1  0.0     1     1.0  1.0  1.0  1.0  0.0   
46  0.0    MSIS           4  1    1  1.0     1     1.0  1.0  0.0  1.0  0.0   
50  0.5     MBA           3  0    0  0.0     1     1.0  1.0  1.0  1.0  1.0   
60  1.0     MBA           3  1    1  1.0     1     1.0  1.0  0.0  1.0  0.0   

    Excel  Tableau  Regression  Classification  Clustering  Bach_0to1  \
23      1        1         0.0               2           2          0   
45      0        1         1.0               2           2          0   
46      1        1         1.0               2           2          0   
50      1        1         1.0               3           3          0   
60      1        0         1.0               1           1          0   

    Bach_1to3  Bach_3to5  Bach_5Plus  Languages  JavaPython1  
23          0          1           0          6            1  
45          1          0           0          7            1  
46          1          0           0          7            1  
50          1          0           0          6            1  
60          0          0           1          7            1

These are the rows of the students who know the highest number of languages within their respective groups (The groups being based on their work status). There are five students total who satisfy this requirement. 

### Question 4

Let us define the "data science experience" of a given person as the person's largest score among Regression, Classification, and Clustering. Compute the average data science experience among all MSIS students. 
(In Camino, pick the correct number)

In [21]:
df['data science experience'] = df[['Regression','Classification','Clustering']].max(axis=1)

In [22]:
df.head()

Job Program  ProgSkills  C  CPP   CS  Java  Python   JS    R  SQL  SAS  \
0  0.0    MSIS           4  1    1  0.0     0     1.0  1.0  0.0  1.0  0.0   
1  0.5    MSIS           3  1    1  0.0     1     0.0  0.0  0.0  1.0  0.0   
2  0.0    MSIS           3  0    0  0.0     1     1.0  0.0  0.0  1.0  0.0   
3  0.0    MSIS           3  1    0  0.0     1     1.0  0.0  1.0  1.0  0.0   
4  0.0    MSIS           3  1    0  0.0     1     1.0  0.0  0.0  1.0  0.0   

   Excel  Tableau  Regression  Classification  Clustering  Bach_0to1  \
0      1        1         1.0               4           4          0   
1      1        0         0.0               2           2          0   
2      1        0         1.0               3           3          0   
3      1        0         1.0               2           3          0   
4      1        0         0.0               1           1          0   

   Bach_1to3  Bach_3to5  Bach_5Plus  Languages  JavaPython1  \
0          1          0           0          6            1   
1          0          0           1          4            1   
2          0          1           0          3            1   
3          0          0           1          5            1   
4          0          1           0          4            1   

   data science experience  
0                      4.0  
1                      2.0  
2                      3.0  
3                      3.0  
4                      1.0

As illustrated above, we have created a new row called 'data science experience' which holds that maximum value between the Regression, Classification, and Clustering columns.

In [23]:
df.groupby('Program')['data science experience'].mean()

Program
Business Man                     3.000
Faculty!                         5.000
MBA                              2.000
MSIS                             1.875
Master of Finance                1.000
Supply Chain Mgmt & Analytics    2.000
Name: data science experience, dtype: float64

We've used groupby to find the average data science experience within each degree program. As illustrated above, the average data science experience for MSIS students is **1.875**.

### Question 5

Among those with at least one year elapsed from their Bachelor's degree, find out who is the "most knowledgeable" person. The most knowledgeable person is the one who knows Classification best (in case of ties, consider whether they know C, then CPP, then CS, then Java, then SAS). Retrieve that person's Program information. (pick the right one in Camino)

In [24]:
df['oneYrSinceBach'] = (df.Bach_1to3==1)|(df.Bach_3to5==1)|(df.Bach_5Plus==1)

In [25]:
df.oneYrSinceBach = df.oneYrSinceBach.astype(int)

In [26]:
df.head(8)

Job                        Program  ProgSkills  C  CPP   CS  Java  Python  \
0  0.0                           MSIS           4  1    1  0.0     0     1.0   
1  0.5                           MSIS           3  1    1  0.0     1     0.0   
2  0.0                           MSIS           3  0    0  0.0     1     1.0   
3  0.0                           MSIS           3  1    0  0.0     1     1.0   
4  0.0                           MSIS           3  1    0  0.0     1     1.0   
5  1.0  Supply Chain Mgmt & Analytics           1  0    0  0.0     0     0.0   
6  0.0                           MSIS           3  1    1  0.0     1     0.0   
7  0.0                           MSIS           2  1    0  0.0     1     0.0   

    JS    R  SQL  SAS  Excel  Tableau  Regression  Classification  Clustering  \
0  1.0  0.0  1.0  0.0      1        1         1.0               4           4   
1  0.0  0.0  1.0  0.0      1        0         0.0               2           2   
2  0.0  0.0  1.0  0.0      1        0         1.0               3           3   
3  0.0  1.0  1.0  0.0      1        0         1.0               2           3   
4  0.0  0.0  1.0  0.0      1        0         0.0               1           1   
5  0.0  0.0  1.0  0.0      1        0         1.0               1           1   
6  0.0  0.0  1.0  0.0      1        0         0.0               2           2   
7  0.0  0.0  1.0  0.0      1        1         0.0               2           2   

   Bach_0to1  Bach_1to3  Bach_3to5  Bach_5Plus  Languages  JavaPython1  \
0          0          1          0           0          6            1   
1          0          0          0           1          4            1   
2          0          0          1           0          3            1   
3          0          0          0           1          5            1   
4          0          0          1           0          4            1   
5          0          1          0           0          1            0   
6          0          1          0           0          4            1   
7          1          0          0           0          3            1   

   data science experience  oneYrSinceBach  
0                      4.0               1  
1                      2.0               1  
2                      3.0               1  
3                      3.0               1  
4                      1.0               1  
5                      1.0               1  
6                      2.0               1  
7                      2.0               0

Illustrated above, we've created a new row, oneYrSinceBach, to display a **1** if its been one year or more since that student acquired a bachelor's degree.

In [27]:
df.groupby('oneYrSinceBach')['Classification'].max()

oneYrSinceBach
0    3
1    4
Name: Classification, dtype: int64

Above, we see the maximum classification scores for both categories (Bachelor's recieved more than a year ago, and bachelor's recieved less than a year ago). We're only concerned with the maximum score for the former category (1).

Let's create a variable storing the value of the maximum classification score for non-recent graduates.

In [28]:
high_class = df.groupby('oneYrSinceBach')['Classification'].max()[1] #Should return 4

In [29]:
df[df.Classification == high_class]

Job   Program  ProgSkills  C  CPP   CS  Java  Python   JS    R  SQL  SAS  \
0   0.0      MSIS           4  1    1  0.0     0     1.0  1.0  0.0  1.0  0.0   
16  1.0  Faculty!           3  1    1  0.0     1     0.0  0.0  0.0  1.0  1.0   
17  1.0       MBA           4  0    0  0.0     0     1.0  NaN  1.0  1.0  NaN   
32  0.0       MBA           3  1    1  0.0     1     1.0  0.0  1.0  1.0  0.0   

    Excel  Tableau  Regression  Classification  Clustering  Bach_0to1  \
0       1        1         1.0               4           4          0   
16      1        1         1.0               4           5          0   
17      1        0         1.0               4           4          0   
32      1        1         1.0               4           4          0   

    Bach_1to3  Bach_3to5  Bach_5Plus  Languages  JavaPython1  \
0           1          0           0          6            1   
16          0          0           1          3            1   
17          0          0           1          3            1   
32          0          0           1          3            1   

    data science experience  oneYrSinceBach  
0                       4.0               1  
16                      5.0               1  
17                      4.0               1  
32                      4.0               1

These are the most experienced students. Let's used a sort_values function to prioritize based on knowledge of C, CPP, CS, Java, and SAS in that order. The highest value among those will be the most experienced student's record.

In [30]:
df[df.Classification == high_class].sort_values(by=['C','CPP','CS','Java','SAS'],ascending = False).head(1)

Job   Program  ProgSkills  C  CPP   CS  Java  Python   JS    R  SQL  SAS  \
16  1.0  Faculty!           3  1    1  0.0     1     0.0  0.0  0.0  1.0  1.0   

    Excel  Tableau  Regression  Classification  Clustering  Bach_0to1  \
16      1        1         1.0               4           5          0   

    Bach_1to3  Bach_3to5  Bach_5Plus  Languages  JavaPython1  \
16          0          0           1          3            1   

    data science experience  oneYrSinceBach  
16                      5.0               1